src/Microsoft.DotNet.DeadRegionAnalysis/AnalysisEngine.Options.cs

```diff
@@ -0,0 +1,200 @@
+                IEnumerable<string> projectPaths = null,
+                IEnumerable<string> sourcePaths = null,
+                IEnumerable<IEnumerable<string>> symbolConfigurations = null,
+                IEnumerable<string> alwaysIgnoredSymbols = null,
+                IEnumerable<string> alwaysDefinedSymbols = null,
+                IEnumerable<string> alwaysDisabledSymbols = null,
+                Tristate undefinedSymbolValue = default(Tristate))
+            {
+                if (projectPaths != null)
+                {
+                    projects = Task.WhenAll(from path in projectPaths select MSBuildWorkspace.Create().OpenProjectAsync(path, CancellationToken.None)).Result;
+                }
+                if (projects != null)
+                {
+                    Documents = GetSharedDocuments(projects);
+                }
+
+                if (sourcePaths != null)
+                {
+                    var projectId = ProjectId.CreateNewId("AnalysisProject");
+                    var solution = new AdhocWorkspace()
+                        .CurrentSolution
+                        .AddProject(projectId, "AnalysisProject", "AnalysisProject", LanguageNames.CSharp);
+
+                    foreach (var path in sourcePaths)
+                    {
+                        var documentId = DocumentId.CreateNewId(projectId);
+                        solution = solution.AddDocument(
+                            documentId,
+                            Path.GetFileName(path),
+                            new FileTextLoader(path, Encoding.UTF8));
+                    }
+
+                    Documents = solution.Projects.Single().Documents;
+                }
+
+                _symbolConfigurations = CalculateSymbolConfigurations(
+                    alwaysDisabledSymbols,
+                    alwaysDefinedSymbols,
+                    alwaysIgnoredSymbols,
+                    symbolConfigurations);
+
+                _undefinedSymbolValue = undefinedSymbolValue;
+            }
+
+            internal CompositePreprocessorExpressionEvaluator GetPreprocessorExpressionEvaluator()
+            {
+                var evaluators = _symbolConfigurations.Select(config => new PreprocessorExpressionEvaluator(config, _undefinedSymbolValue));
+                return new CompositePreprocessorExpressionEvaluator(evaluators);
+            }
+
+            internal PreprocessorSymbolTracker GetPreprocessorSymbolTracker()
+            {
+                var specifiedSymbols = new HashSet<string>();
+
+                foreach (var config in _symbolConfigurations)
+                {
+                    foreach (string symbol in config.Keys)
+                    {
+                        specifiedSymbols.Add(symbol);
+                    }
+                }
+
+                return new PreprocessorSymbolTracker(specifiedSymbols);
+            }
+
+            private static IEnumerable<Document> GetSharedDocuments(IEnumerable<Project> projects)
```

Comment:
Sounds good.

As a note to myself: I don't think I've documented the fact that we take the intersection of documents in the specified projects when multiple projects are passed on the command line.  Originally, I passed two projects which have different build configurations, but many shared files on the command line, and intersected the state of all the conditional regions in each project.

Since updating the algorithm, the build configurations specified by the projects themselves are ignored because whichever configurations msbuild deduces from the project files is dependent on the environment.  It's clearer to specify each configuration explicitly on the command line, an easy way to get each configuration being to look at the csc commands run by msbuild.